### PR TITLE
ENH: Distance dependence measures

### PR DESCRIPTION
**Pull request by roncho12**
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

An implementation of Distance dependence measures and the Distance Covariance (dCov) test, as introduced by Szekely, G.J., Rizzo, M.L., and Bakirov, N.K. (2007) “Measuring and testing dependence by correlation of distances”. Annals of Statistics, Vol. 35 No. 6, pp. 2769-2794.

Any and all feedback is most welcome and I'd be happy to make any changes if need be.

### CODE REVIEW

File: statsmodels/stats/dist_dependence_measures.py [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+    x_dist = squareform(pdist(x, "euclidean"))
+
+    for i in range(B):
+        np.random.shuffle(y)
+        emp_dist[i] = calc_statistics(x, y, x_dist=x_dist).test_statistic
+
+    return emp_dist
+
+
+def calc_statistics(x, y, x_dist=None, y_dist=None):
```
Comment (bashtage) [inline statsmodels/stats/dist_dependence_measures.py]: Good looking PR.  The only quibble I have is about naming.  The function names should be as clear as possible without referencing the module name.  Perhaps `distance_statistics` here (open to suggestions).

File: statsmodels/stats/dist_dependence_measures.py [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+        test_statistic=test_statistic,
+        dcor=dcor,
+        dcov=dcov,
+        dvar_x=dvar_x,
+        dvar_y=dvar_y,
+        S=S,
+    )
+
+
+def dcov(x, y):
```
Comment (bashtage) [inline statsmodels/stats/dist_dependence_measures.py]: `covariance_distance`?  I assume the paper calls it dCov.

Comment (bashtage) [reply statsmodels/stats/dist_dependence_measures.py]: Or maybe `distance_covariance`

File: statsmodels/stats/dist_dependence_measures.py [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+
+    >>> from statsmodels.stats.dist_dependence_measures import dcov
+    >>> dcov(np.random.random(1000), np.random.random(1000))
+    0.007575063951951362
+
+    """
+    return calc_statistics(x, y).dcov
+
+
+def dvar(x):
```
Comment (bashtage) [inline statsmodels/stats/dist_dependence_measures.py]: distance_variance?

File: statsmodels/stats/dist_dependence_measures.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+
+    >>> from statsmodels.stats.dist_dependence_measures import dvar
+    >>> dvar(np.random.random(1000))
+    0.21732609190659702
+
+    """
+    return dcov(x, x)
+
+
+def dcor(x, y):
```
Comment (bashtage) [inline statsmodels/stats/dist_dependence_measures.py]: distance_correlation

**Comment by roncho12** [comment]
@bashtage Thanks!
I've changed the names.
You're correct, i used the same naming as the paper, but it does make sense to make them more explicit.

File: statsmodels/stats/dist_dependence_measures.py (hunk 1 above)
Comment (roncho12) [reply statsmodels/stats/dist_dependence_measures.py]: changed

File: statsmodels/stats/dist_dependence_measures.py (hunk 2 above)
Comment (roncho12) [reply statsmodels/stats/dist_dependence_measures.py]: changed

File: statsmodels/stats/dist_dependence_measures.py (hunk 4 above)
Comment (roncho12) [reply statsmodels/stats/dist_dependence_measures.py]: changed

File: statsmodels/stats/dist_dependence_measures.py [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+
+
+DistDependStat = namedtuple(
+    "DistDependStat",
+    ["test_statistic", "distance_correlation",
+     "distance_covariance", "dvar_x", "dvar_y", "S"],
+)
+
+
+def dcovtest(x, y, B=None, method="auto"):
```
Comment (bashtage) [inline statsmodels/stats/dist_dependence_measures.py]: I missed one, sorry.  Probably distance_covariance_test, or perhaps test_distance_covariance.  Your choice. Then ready after green.

Comment (roncho12) [reply statsmodels/stats/dist_dependence_measures.py]: missed it too :)
fixed

**Comment by bashtage** [comment]
Thanks.

**Pull request merged**